client/txn.go

```diff
@@ -384,47 +378,40 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) error {
 // always commit or clean-up explicitly even when that may not be
 // required (or even erroneous).
 func (txn *Txn) send(calls ...proto.Call) error {
+	if txn.Proto.Status != proto.PENDING {
+		return util.Errorf("attempting to use %s transaction", txn.Proto.Status)
+	}
+
 	lastIndex := len(calls) - 1
 	if lastIndex < 0 {
 		return nil
 	}
 
+	lastReq := calls[lastIndex].Args
 	// haveTxnWrite tracks intention to write. This is in contrast to
 	// txn.Proto.Writing, which is set by the coordinator when the first
 	// intent has been created, and which lives for the life of the
 	// transaction.
-	var haveTxnWrite bool
-	var haveEndTxn bool
+	haveTxnWrite := proto.IsTransactionWrite(lastReq)
 
-	for i, call := range calls {
+	for _, call := range calls[:lastIndex] {
```

Comment:
This is subtle; I had to read it a few times to figure out why we chop off the last call. I think the old version which initialized `haveTxnWrite` in this loop and checked `i != lastIndex` was easier to follow. If we keep this structure, it should have a comment like "We only care about `haveTxnWrite` because it allows us to chop off an unnecessary `EndTransactionRequest`. Therefore the write status of the last call never matters, only whether it is an `EndTransactionRequest`."

Separately, this logic has always struck me as an odd optimization. It seems to me that TxnCoordSender should just send the EndTransaction call whether it's necessary or not, and the decision to elide EndTransaction calls for read-only transactions should be made at a higher level in the client.